contracts/core/mailbox/src/execute.rs

```diff
@@ -0,0 +1,598 @@
+        config.default_ism = Some(new_default_ism);
+
+        Ok(config)
+    })?;
+
+    Ok(Response::new().add_event(event))
+}
+
+pub fn set_default_hook(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_default_hook: String,
+) -> Result<Response, ContractError> {
+    ensure_eq!(
+        get_owner(deps.storage)?,
+        info.sender,
+        ContractError::Unauthorized {}
+    );
+
+    let new_default_hook = deps.api.addr_validate(&new_default_hook)?;
+    let event = emit_default_hook_set(info.sender, new_default_hook.clone());
+
+    CONFIG.update::<_, ContractError>(deps.storage, |mut config| {
+        config.default_hook = Some(new_default_hook);
+
+        Ok(config)
+    })?;
+
+    Ok(Response::new().add_event(event))
+}
+
+pub fn set_required_hook(
+    deps: DepsMut,
+    info: MessageInfo,
+    new_required_hook: String,
+) -> Result<Response, ContractError> {
+    ensure_eq!(
+        get_owner(deps.storage)?,
+        info.sender,
+        ContractError::Unauthorized {}
+    );
+
+    let new_required_hook = deps.api.addr_validate(&new_required_hook)?;
+    let event = emit_required_hook_set(info.sender, new_required_hook.clone());
+
+    CONFIG.update::<_, ContractError>(deps.storage, |mut config| {
+        config.required_hook = Some(new_required_hook);
+
+        Ok(config)
+    })?;
+
+    Ok(Response::new().add_event(event))
+}
+
+pub fn dispatch(
+    deps: DepsMut,
+    info: MessageInfo,
+    dispatch_msg: DispatchMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    let nonce = NONCE.load(deps.storage)?;
+
+    ensure!(
+        dispatch_msg.recipient_addr.len() <= 32,
+        ContractError::InvalidAddressLength {
+            len: dispatch_msg.recipient_addr.len()
+        }
+    );
+
+    // calculate gas
+    let required_hook = config.required_hook.expect("required_hook not set");
+    let (received_value, required_value) = get_required_value(
+        &deps.querier,
+        &info,
+        required_hook.as_str(),
+        dispatch_msg.metadata.clone().unwrap_or_default(),
+        dispatch_msg.msg_body.clone(),
+    )?;
+
+    // interaction
+    let hook = dispatch_msg
+        .get_hook_addr(deps.api, config.default_hook)?
+        .expect("default_hook not set");
+    let hook_metadata = dispatch_msg.metadata.clone().unwrap_or_default();
+
+    let msg = dispatch_msg.to_msg(MAILBOX_VERSION, nonce, config.local_domain, &info.sender)?;
+    let msg_id = msg.id();
+
+    // effects
+    NONCE.save(deps.storage, &(nonce + 1))?;
+    LATEST_DISPATCHED_ID.save(deps.storage, &msg_id.to_vec())?;
+
+    ensure!(
+        received_value.amount >= required_value.amount,
+        ContractError::InsufficientFunds {
+            required: required_value,
+            received: received_value,
+        }
+    );
+
+    // make message
+    let post_dispatch_msgs = vec![
+        post_dispatch(
+            required_hook,
```

Comment:
```suggestion
            msg,
```
do we need to clone twice?